pkg/operator/ceph/cluster/cephx.go

```diff
@@ -0,0 +1,457 @@
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/ceph/cluster/mon"
+	"github.com/rook/rook/pkg/operator/ceph/config"
+	"github.com/rook/rook/pkg/operator/ceph/config/keyring"
+	"github.com/rook/rook/pkg/operator/ceph/controller"
+	"github.com/rook/rook/pkg/operator/ceph/reporting"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	//#nosec G101 -- This is only a secret name prefix
+	adminRotatorSecretPrefix = "rook-ceph-admin-rotator"
+	//#nosec G101 -- This is only the admin client name
+	adminRotatorUsername = "client.admin-rotator"
+)
+
+// note: must match adminKeyringTemplate
+var (
+	adminKeyAccessCaps = []string{"mds", "allow *", "mon", "allow *", "osd", "allow *", "mgr", "allow *"}
+
+	// ReloadManager sends SIGHUP to the main process. Allow this to be stubbed for unit tests.
+	reloadManagerFunc = controller.ReloadManager
+
+	// admin key rotation ends with a rook restart, so return a reconcile error after to help
+	// prevent rook from continuing reconcile when restart is imminent
+	errSuccessfulAdminKeyRotation = fmt.Errorf("successful admin cephx key rotation requires the current cluster reconcile to restart")
+
+	// Prevent multiple admin rotation operations from running simultaneously. This is important
+	// because the admin rotation routine does not stop as soon as the parent context is canceled.
+	// in the worst case, it could still be running while a CephCluster re-reconcile gets back
+	// around to admin key rotation again.
+	adminRotationMutex      = sync.Mutex{}
+	adminRotationInProgress = map[string]struct{}{}
+)
+
+// claimAdminRotationLock tries to claim a lock on admin key rotation for a cluster namespace
+// returns error if the lock could not be granted.
+// Because (*Mutex).Lock() blocks while trying to get a lock, we could end up stalling Rook
+// reconciles if we rely solely on the lock itself. To prevent stalls, use a map 'set' to indicate
+// whether rotation is in progress, and only lock access to the set. When a rotation is locked,
+// return an error so that the reconcile can be requeued to try again later.
+func claimAdminRotationLock(namespace string) error {
+	adminRotationMutex.Lock()
+	defer adminRotationMutex.Unlock()
+	if _, ok := adminRotationInProgress[namespace]; ok {
+		return fmt.Errorf("failed to claim admin rotation lock for cluster in namespace %q", namespace)
+	}
+	logger.Debugf("claimed admin rotation lock for cluster in namespace %q", namespace)
+	adminRotationInProgress[namespace] = struct{}{}
+	return nil
+}
+
+// releaseAdminRotationLock releases a claimed lock for a cluster namespace
+func releaseAdminRotationLock(namespace string) {
+	adminRotationMutex.Lock()
+	defer adminRotationMutex.Unlock()
+	delete(adminRotationInProgress, namespace)
+}
```

Comment:
@travisn any concerns with this mutex/lock setup?

Reply:
LGTM